### ISSUE TITLE
Framing + @vocab Bug (playground or algorithm?)

### ISSUE THREAD

**Issue by azaroth42** [json-ld/json-ld.org#404]
Issue: 
When providing a frame for content that uses @vocab to map from strings to full URIs (per http://www.w3.org/TR/json-ld/#type-coercion), the resulting URI is ... basically ... nonsense.  The other views get the correct URI.

Example: http://tinyurl.com/ppsnlte

From: https://github.com/IIIF/iiif.io/issues/623 and https://github.com/IIIF/iiif.io/issues/626

**Comment by lanthaler** [comment]
Could you please post an example (including the context/frame) of what you see and what you would expect to see.

**Comment by azaroth42** [comment]
They're in the example provided via permalink to the playground, but:

Frame:
```
{
  "@context": "http://iiif.io/api/image/2/context.json",
  "@embed": "true",
  "protocol": "http://iiif.io/api/image/"
}
```

Data:

```
{
  "@context": {
    "iiif": "http://iiif.io/api/image/2#",
    "xsd": "http://www.w3.org/2001/XMLSchema#",
    "doap": "http://usefulinc.com/ns/doap#",
    "cors": {
      "@id": "iiif:corsFeature",
      "@type": "iiif:Feature"
    },
    "sizeByW": {
      "@id": "iiif:sizeByWFeature",
      "@type": "iiif:Feature"
    },
    "canonicalLinkHeader": {
      "@id": "iiif:canonicalLinkHeaderFeature",
      "@type": "iiif:Feature"
    },
    "protocol": {
      "@id": "doap:implements",
      "@type": "@id"
    },
    "supports": {
      "@id": "iiif:supports",
      "@type": "@vocab"
    }
  },
  "protocol": "http://iiif.io/api/image/",
  "supports": [
    "canonicalLinkHeader",
    "sizeByW"
  ]
}
```

Results in:

```
{
  "@context": "http://iiif.io/api/image/2/context.json",
  "@graph": [
    {
      "@id": "_:b0",
      "supports": [
        "iiif:canonicalLinkHeaderFeature",
        "sizes:ByWFeature"
      ],
      "protocol": "http://iiif.io/api/image/"
    }
  ]
}
```

Whereas the result should be:

```
{
  "@context": "http://iiif.io/api/image/2/context.json",
  "@graph": [
    {
      "@id": "_:b0",
      "supports": [
        "iiif:canonicalLinkHeaderFeature",
        "iiif:sizeByWFeature"
      ],
      "protocol": "http://iiif.io/api/image/"
    }
  ]
}
```

**Comment by gkellogg** [comment]
I believe you’re running afoul of the _compactIri_ algorithm which picks the shortest resulting compact IRI, in this case using the `sizes` term as a prefix. I found this anoying, and made a change to my own implementation which favors terms which are defined without an object body, in this case `iiid`, but that is not part of the standard.

What’s happening here is that `iiid:sizeByWFeature` is expanded to `http://iiif.io/api/image/2#:sizeByWFeature`. However, when compacting, the `sizes` term is selected, as `sizes:ByWFeature` is shorter than `iiid:sizeByWFeature`.

This works properly according to the spec, but often results in compact IRIs that people aren’t looking for (schema.org has similar issues).

In my case, I prefer prefixes which are defined like `"iiif" :   "http://iiif.io/api/image/2#”` over `"iiif" :   {“@id”: "http://iiif.io/api/image/2#”}`, as I believe this generates better output, but that is outside of the spec. Something which should be addressed in a 1.1 version of the specs.

Note that this isn’t a framing issue, but a compaction issue.

**Comment by lanthaler** [comment]
Yeah, unfortunately there isn't much you can do about this as we don't differentiate between terms that should/can be use as prefixes and terms that shouldn't.

**Comment by azaroth42** [comment]
Ahh, rats :(

Thanks for the explanation!

**Comment by azaroth42** [comment]
@gkellogg Is there an issue specifically about the compaction algorithm that I could +1? Or will this issue serve?  If so, shall we change the title?

**Comment by gkellogg** [comment]
Closed in favor of #405.
